include/tvm/ir.h

```diff
@@ -16,36 +16,581 @@
+};
+
+/*!
+ * \brief return true_value if condition is true, otherwise return false_value.
+ * \note Both true_value and false_value could be evaluated
+ *       regardless of the condition value.
+ *       Do not use it to guard against out of bound access,
+ *       please use if_then_else instead.
+ */
+class Select : public ExprNode {
+ public:
+  /*! \brief The condition */
+  Expr condition;
+  /*! \brief value to be returned when condition is true. */
+  Expr true_value;
+  /*! \brief value to be returned when condition is false. */
+  Expr false_value;
+
+  void VisitAttrs(AttrVisitor* v) final {
+    v->Visit("dtype", &type);
+    v->Visit("condition", &condition);
+    v->Visit("true_value", &true_value);
+    v->Visit("false_value", &false_value);
+  }
+
+  TVM_DLL static Expr make(Expr condition, Expr true_value, Expr false_value);
+
+  static constexpr const char* _type_key = "Select";
+  TVM_DECLARE_NODE_TYPE_INFO(Select, ExprNode);
+};
+
+/*!
+ * \brief Load the value from buffer_var.
+ *
+ *  Equivalent to ((DType*)buffer_var)[index]
+ *  where DType is the type specified by type().element_of().
+ *
+ *  For example, if type = float32x3, then the load will corresponds to
+ *
+ * \code
+ *
+ *  auto buffer = static_cast<float*>(buffer_var);
+ *  auto loaded_val = float32x3(buffer[index.v0], buffer[index.v1], buffer[index.v2]);
+ *
+ * \endcode
+ */
+class Load : public ExprNode {
+ public:
+  /*! \brief The buffer variable. */
+  Var buffer_var;
+  /*! \brief The index locations to be loaded. */
+  Expr index;
+  /*! \brief The predicate to mask which lanes would be loaded. */
+  Expr predicate;
+
+  void VisitAttrs(AttrVisitor* v) final {
+    v->Visit("dtype", &type);
+    v->Visit("buffer_var", &buffer_var);
+    v->Visit("index", &index);
+    v->Visit("predicate", &predicate);
+  }
+
+  TVM_DLL static Expr make(Type type, Var buffer_var, Expr index, Expr predicate);
+
+  static constexpr const char* _type_key = "Load";
+  TVM_DECLARE_NODE_TYPE_INFO(Load, ExprNode);
+};
+
+/*!
+ * \brief Construct a vector with lanes elements
+ *        where its i-th element equals base + i * stride.
+ *  This is useful to construct a index for a continuous vector load.
+ *
+ *  Examples:
+ *  - ramp(0, 1, 3) = [0, 1, 2]
+ *  - ramp(1, 2, 4) = [1, 3, 5, 7]
+ */
+class Ramp : public ExprNode {
+ public:
+  /*! \brief The base value. */
+  Expr base;
+  /*! \brief The stride of each step. */
+  Expr stride;
+  /*! \brief Total number of lanes. */
+  int lanes;
+
+  void VisitAttrs(AttrVisitor* v) final {
+    v->Visit("dtype", &type);
+    v->Visit("base", &base);
+    v->Visit("stride", &stride);
+    v->Visit("lanes", &lanes);
+  }
+
+  TVM_DLL static Expr make(Expr base, Expr stride, int lanes);
+
+  static constexpr const char* _type_key = "Ramp";
+  TVM_DECLARE_NODE_TYPE_INFO(Ramp, ExprNode);
+};
+
+/*! \brief Create a vector where all the elements are value. */
+class Broadcast : public ExprNode {
```

Comment:
This might be out of the scope for now, but I think we should use the `XNode` and `X` for the Node and NodeRef subclasses, it might be a good time to stop using raw pointers in the functor as well